### PR TITLE
Version fix

### PR DESCRIPTION
**Pull request by RoiEXLab**
This fixes the last noteable regression.
Note that my assumption is that as long the major and minor version number are the same, a release is considered compatible.
I did some testing and it turns out this change works without requiring to update bots as long as the third (point) version number isn't changed.
If this version number is changed, the bots need to be updated to consider the client as compatible.

I also did a lot of cleanup, the core change is very small (just view the first 3 commits)

### CODE REVIEW

**Comment by simon33-2** [comment]
I believe it should be the first 3 digits are checked - for consistency. But since we have remove old jar and the current version has a zero in the 3rd digit, this should work fine.

File: src/main/java/games/strategy/util/Version.java [hunk 1]
```diff
@@ -110,33 +133,17 @@ public boolean equals(final Object o) {
...
-    return compareTo(o, false);
-  }
-
-  public int compareTo(final Version other) {
-    return compareTo(other, false);
-  }
-
-  private int compareTo(final Object o, final boolean ignoreMicro) {
-    if (o == null) {
+    if (o == null || !(o instanceof Version)) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/util/Version.java]: Nit: `o instanceof Version` will return `false` if `o` is `null`, so the `o == null` check is redundant.

File: src/main/java/games/strategy/util/Version.java [hunk 2]
```diff
@@ -167,29 +174,45 @@ private int compareTo(final Version other, final boolean ignoreMicro) {
...
   }
 
-  private String toStringFull(final String separator, final boolean noMicro) {
-    return m_major + separator + m_minor + separator + m_point
-      + (noMicro ? "" : (separator + (m_micro == Integer.MAX_VALUE ? "dev" : m_micro)));
+  /**
+   * Creates a complete version string, even if some version numbers are 0.
+   */
+  public String toStringFull() {
+    return Joiner.on('.').join(m_major, m_minor, m_point, m_micro == Integer.MAX_VALUE ? "dev" : m_micro);
```
Comment (ssoloff) [inline src/main/java/games/strategy/util/Version.java]: Will `java.lang.String#join()` or `java.util.StringJoiner` not work here?  Just trying to avoid the Guava dependency if the functionality is available in the JDK.

File: src/test/java/games/strategy/util/VersionTest.java [hunk 3]
```diff
@@ -74,4 +65,12 @@ public void getExactVersion() {
     assertEquals("1.2.3.4", new Version(1, 2, 3, 4).getExactVersion());
     assertEquals("1.2.3.4.5", new Version("1.2.3.4.5").getExactVersion());
   }
+
+  @Test
+  public void testCompatible() {
+    assertTrue(new Version(1, 9).isCompatible(new Version(1, 9)));
+    assertFalse(new Version(2, 9).isCompatible(new Version(1, 9)));
+    assertFalse(new Version(1, 10).isCompatible(new Version(1, 9)));
+    assertFalse(new Version(2, 10).isCompatible(new Version(1, 9)));
```
Comment (ssoloff) [inline src/test/java/games/strategy/util/VersionTest.java]: Would it be appropriate to add some tests that demonstrate `m_point` and `m_micro` are not considered by the algorithm?  For example (untested):

```java
assertTrue(new Version(1, 9, 0).isCompatible(new Version(1, 9, 1)));
```

Comment (RoiEXLab) [reply src/test/java/games/strategy/util/VersionTest.java]: Good Point

File: src/main/java/games/strategy/util/Version.java (hunk 2 above)
Comment (RoiEXLab) [reply src/main/java/games/strategy/util/Version.java]: In theory yes, but this would require us to put a `String.valueOf` around every var

**Comment by RoiEXLab** [comment]
@ssoloff I changed the assertTrue calls to assertEquals before reading your message ^^

Updated!

**Comment by RoiEXLab** [comment]
Also @ssoloff do you think we should consider a release incompatible when any of the 3 "major/minor/point" version numbers are changed as @simon33-2 suggested?

**Comment by ssoloff** [comment]
@RoiEXLab 

> do you think we should consider a release incompatible when any of the 3 "major/minor/point" version numbers are changed as @simon33-2 suggested?

As I said in #1768, I still don't really understand what our current version compatibility rules are. :smile:  If we followed semver, obviously only the major version would have to be checked.  @simon33-2 seems to have a good understanding of what's required with respect to the current compatibility matrix, so I think implementing his suggestion would be ideal, as he said, for consistency, even if the current solution just happens to work because the two versions under consideration have a point version of zero.

File: src/main/java/games/strategy/util/Version.java (hunk 2 above)
Comment (ssoloff) [reply src/main/java/games/strategy/util/Version.java]: Ah, yes.  I didn't bother to look at the signatures to see that they both require a `CharSequence`.

File: src/main/java/games/strategy/util/Version.java [hunk 4]
```diff
@@ -167,29 +174,45 @@ private int compareTo(final Version other, final boolean ignoreMicro) {
...
   @Override
   public String toString() {
-    return m_point == 0 && m_micro == 0 ? m_major + "." + m_minor : toStringFull(".", m_micro == 0);
+    return m_micro != 0 ? toStringFull() : m_major + "." + m_minor + (m_point != 0 ? "." + m_point : "");
+  }
+
+  /**
+   * Checks if The major and minor version numbers are equal.
+   * Returns true if they are.
+   */
```
Comment (ssoloff) [inline src/main/java/games/strategy/util/Version.java]: This Javadoc is no longer accurate after the recent change that included the point version in the comparison.

File: src/test/java/games/strategy/util/VersionTest.java [hunk 5]
```diff
@@ -12,66 +13,70 @@
...
 
   @Test
   public void getExactVersion() {
     assertEquals("1.2.3.4", new Version(1, 2, 3, 4).getExactVersion());
     assertEquals("1.2.3.4.5", new Version("1.2.3.4.5").getExactVersion());
   }
+
+  @Test
+  public void testCompatible() {
+    assertTrue(new Version(1, 9).isCompatible(new Version(1, 9)));
```
Comment (DanVanAtta) [inline src/test/java/games/strategy/util/VersionTest.java]: - recommend ordering the test values, this is something that is useful in combinatorics when you want to carefully enumerate all permutations, eg:
```
1.9 + 1.9
1.9 + 1.9.0
1.9 + 1.9.1
1.9 + 1.9.9
1.9 + 1.9.0.0
1.9 + 1.9.0.1
1.9 + 1.9.0.9
```

Second, Version can be serialized, if there is not a match then the fields or object itself may be null, so another interesting test case would be `null`

**Comment by DanVanAtta** [comment]
NB: the (class javadoc on Version.java )[https://github.com/triplea-game/triplea/blob/master/src/main/java/games/strategy/util/Version.java#L9]
 defines compatibility.

NB: We had this rough plan as well for next version numbers: https://github.com/triplea-game/triplea/issues/323#issuecomment-170134307

---------------------

@ssoloff I could not repro, the 'remember me' function worked as well for me.

**Comment by DanVanAtta** [comment]
Testing was fine here. @ssoloff if problem is persisent let us know.

**Comment by ssoloff** [comment]
@DanVanAtta 

> I could not repro, the 'remember me' function worked as well for me.

If you just run from the IDE, it will work because the client version is using "1.9.0" read from `game_engine.properties`.  Try changing that to a value similar to what the Gradle build would produce, e.g. "1.9.0.7777".  In that case, I get the `CouldNotLoginException`.

Since you merged, I'll try to repro with the installer once the build is done.

BTW, is _"the 'remember me' function worked as well for me"_ related to this PR?  Or are you talking about something else?

**Comment by ssoloff** [comment]
Confirmed I can repro the problem in 1.9.0.7040.

So I think the root cause is that the `Version#isCompatible()` call that's causing authentication to fail is running in the server (i.e. the bot).  See `ClientLoginValidator#verifyConnection()`.  It's using the legacy definition of "compatibility" from 3635.  This might require a bot upgrade to get it to work.

**Comment by ssoloff** [comment]
This should be obvious from the previous comment, but I confirmed that a 7040 client cannot connect to a 3635 server in a peer-to-peer network game either.  That is, this issue is not restricted to just bots.  Or more succinctly, we have a compatibility issue. :smile:  I don't believe upgrading the bots will be sufficient.  We may have to revisit #2394 before releasing.

I think we (devs) didn't detect this in local testing because we typically run the client from the IDE.  As I said above, the version in `game_engine.properties` leaves the micro version unspecified, whereas this is not the case for the Travis build.  Again, the easy way to repro this is to set the micro version in `game_engine.properties`, e.g. `1.9.0.7040`, before running the client from the IDE.

**Comment by prastle** [comment]
I can also confirm error still exists when joining my own host or a bot With the just released 20 min ago.
@ssoloff

**Comment by DanVanAtta** [comment]
Looks like I'm the oddball in testing, that is regrettable. It's time to start rolling back then, #2470 and #2394

**Comment by DanVanAtta** [comment]
Revert hammer came out, let's check out the next build, it should have the breaking updates removed

**Comment by ssoloff** [comment]
:+1: Confirmed I could connect to a 3635 bot from a 7051 client.

**Comment by prastle** [comment]
.49 could connect as well but not myself

File: src/main/java/games/strategy/engine/ClientFileSystemHelper.java [hunk 6]
```diff
@@ -67,7 +67,7 @@ private static String getGameRunnerFileLocation(final String runnerClassName) {
 
   private static String getTripleaJarWithEngineVersionStringPath() {
     final Version version = ClientContext.engineVersion();
-    return "triplea_" + version.toStringFull("_") + ".jar!";
+    return "triplea_" + version.toStringFull().replaceAll("\\.", "_") + ".jar!";
```
[the remark this answers is not in the document]
Comment (simon33-2) [reply src/main/java/games/strategy/engine/ClientFileSystemHelper.java]: I agree. The prior form makes more sense. The toStringFull() method just calls toStringFull("\\.") so why do replaceAll?
